subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/transaction/CompileTransaction.java

```diff
@@ -0,0 +1,302 @@
+                .forEach(path -> path.toFile().delete());
+        }
+    }
+
+    private boolean isDirectoryAlsoInOtherRoot(Path directory, Path root, File otherRoot) {
+        Path relativePath = root.relativize(directory);
+        File fileInOtherRoot = new File(otherRoot, relativePath.toString());
+        return fileInOtherRoot.isDirectory();
+    }
+
+    private void deleteRecursively(File file) {
+        try {
+            deleter.deleteRecursively(file);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private List<StagedOutput> collectOutputsToStage() {
+        List<StagedOutput> stagedOutputs = new ArrayList<>();
+        stagedOutputs.add(new StagedOutput(spec.getDestinationDir(), new File(tempDir, "compile-output"), spec::setDestinationDir));
+
+        File annotationOutputDir = spec.getCompileOptions().getAnnotationProcessorGeneratedSourcesDirectory();
+        if (annotationOutputDir != null) {
+            StagedOutput stagedOutput = new StagedOutput(annotationOutputDir, new File(tempDir, "annotation-output"), file -> spec.getCompileOptions().setAnnotationProcessorGeneratedSourcesDirectory(file));
+            stagedOutputs.add(stagedOutput);
+        }
+
+        File headerOutputDir = spec.getCompileOptions().getHeaderOutputDirectory();
+        if (spec.getCompileOptions().getHeaderOutputDirectory() != null) {
+            StagedOutput stagedOutput = new StagedOutput(headerOutputDir, new File(tempDir, "header-output"), file -> spec.getCompileOptions().setHeaderOutputDirectory(file));
+            stagedOutputs.add(stagedOutput);
+        }
+        return stagedOutputs;
+    }
+
+    private StashResult stashFilesThatShouldBeDeleted() {
+        int uniqueId = 0;
+        File compileOutput = spec.getDestinationDir();
+        File annotationProcessorOutput = spec.getCompileOptions().getAnnotationProcessorGeneratedSourcesDirectory();
+        File headerOutput = spec.getCompileOptions().getHeaderOutputDirectory();
+        List<StashedFile> stashedFiles = new ArrayList<>();
+        for (File sourceFile : collectFilesToStash(compileOutput, annotationProcessorOutput, headerOutput)) {
+            File stashedFile = new File(stashDirectory, sourceFile.getName() + ".uniqueId" + uniqueId++);
+            moveFile(sourceFile, stashedFile);
+            stashedFiles.add(new StashedFile(sourceFile, stashedFile));
+        }
+        List<File> sourceDirectories = Stream.of(compileOutput, annotationProcessorOutput, headerOutput)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        return new StashResult(sourceDirectories, stashedFiles);
+    }
+
+    private Set<File> collectFilesToStash(File compileOutput, @Nullable File annotationProcessorOutput, @Nullable File headerOutput) {
+        Set<File> filesToStash = new HashSet<>();
```

Comment:
Thanks for pointing this out. Rethinking this, I came to the conclussion that order is not important, since stashed files are not modified and one file does not affect another file (files are not connected).